verilog/formatting/tree_unwrapper.cc

```diff
@@ -1361,6 +1459,43 @@ void TreeUnwrapper::ReshapeTokenPartitions(
         break;
       }
 
+    case NodeEnum::kParamDeclaration: {
+      {
+        const auto* trailing_assign = GetSubtreeAsSymbol(
+            node, NodeEnum::kParamDeclaration, 2);
+        if (trailing_assign) {
+          const auto& node = SymbolCastToNode(*trailing_assign);
+          if (NodeEnum(node.Tag().tag) == NodeEnum::kTrailingAssign) {
+            const auto& expression = GetSubtreeAsNode(
+                node, NodeEnum::kTrailingAssign, 1);
+            if (expression.children().size() == 1) {
+              const auto& expr = GetSubtreeAsNode(
+                  expression, NodeEnum::kExpression, 0);
+              switch (NodeEnum(expr.Tag().tag)) {
+                case NodeEnum::kRandomizeMethodCallExtension:
+                case NodeEnum::kRandomizeFunctionCall:
+                case NodeEnum::kFunctionCall:
+                case NodeEnum::kSystemTFCall: {
```

Comment:
Putting this logic here doesn't seem to cover general cases of expressions adequately.  The way I would think of this is:

* How should a standalone call/expression be shaped?  (without regard to parent context)
* How should the result of that shaping be integrated at the next level up, and what _incremental_ changes are needed to do the right thing?  (e.g. merging partitions, adjusting wrapping/indentation of sub-partitions).  Example contexts:
  * In statements
  * In initialization/defaults of parameters
  * On right-sides of assignments
  * Deeper inside other expressions

This will eliminate duplication and minimize deep special casing.

Reply:
Statements and right-side assignments reuse current code.
Parameters (and other parents) are handled here https://github.com/google/verible/pull/246/files#diff-7872d97b05fbc6f2a64cf14b8ae74d75R863

I don't know how nested calls and if-statements should be handled.
For example if-statement can have binary expression, e.g. ```a + b(c, d, e)``` which currently is partitioned like:
```
      { (>>[<auto>], policy: append-fitting-sub-partitions) @{25,1,0}, (origin: "if (a + b(c,d,e));")
        { (>>[if ( a + b (], policy: fit-else-expand) }
        { (>>[<auto>], policy: always-expand) @{25,1,0,1}, (origin: "c,d,e")
          { (>>[c ,], policy: fit-else-expand, (origin: "c")) }
          { (>>[d ,], policy: fit-else-expand, (origin: "d")) }
          { (>>[e ) ) ;], policy: fit-else-expand, (origin: "e")) }
        }
      }
```
but such statement ```b(c,d,e) + a``` is partitioned like:
```
      { (>>[<auto>], policy: fit-else-expand) @{25,1,1}, (origin: "if (b(c,d,e)+c);")
        { (>>[<auto>], policy: append-fitting-sub-partitions) @{25,1,1,0}, (origin: "b(c,d,e)")
          { (>>[if ( b (], policy: fit-else-expand) }
          { (>>[<auto>], policy: always-expand) @{25,1,1,0,1}, (origin: "c,d,e")
            { (>>[c ,], policy: fit-else-expand, (origin: "c")) }
            { (>>[d ,], policy: fit-else-expand, (origin: "d")) }
            { (>>[e )], policy: fit-else-expand, (origin: "e")) }
          }
        }
        { (>>[+ c ) ;], policy: uninitialized) }
      }
```
I guess the ```+ c );``` should appended to last partition: ```e )```. But what about such expression:
```
      { (>>[<auto>], policy: append-fitting-sub-partitions) @{25,1,0}, (origin: "a(b,c)")
        { (>>[if ( a (], policy: fit-else-expand) }
        { (>>[<auto>], policy: always-expand) @{25,1,0,1}, (origin: "b,c")
          { (>>[b ,], policy: fit-else-expand, (origin: "b")) }
          { (>>[c )], policy: fit-else-expand, (origin: "c")) }
        }
      }
      { (>>[+], policy: uninitialized) }
      { (>>[<auto>], policy: append-fitting-sub-partitions) @{25,1,2}, (origin: "d(e,f)")
        { (>>[d (], policy: fit-else-expand) }
        { (>>[<auto>], policy: always-expand) @{25,1,2,1}, (origin: "e,f")
          { (>>[e ,], policy: fit-else-expand, (origin: "e")) }
          { (>>[f ) ) ;], policy: fit-else-expand, (origin: "f")) }
        }
      }
```
Should the ```+``` partition be appended to previous partition or prepended to next partition?
Or should the whole expression be flatten? I mean treat ```+``` as comma and ```if ()``` as function call?

Reply:
Yes, `b(c,d,e) + a` should probably contain `+ c)` up one level like this:

```
      { (>>[<auto>], policy: fit-else-expand) @{25,1,1}, (origin: "if (b(c,d,e)+c);")
        { (>>[<auto>], policy: append-fitting-sub-partitions) @{25,1,1,0}, (origin: "b(c,d,e)")
          { (>>[if ( b (], policy: fit-else-expand) }
          { (>>[<auto>], policy: always-expand) @{25,1,1,0,1}, (origin: "c,d,e")
            { (>>[c ,], policy: fit-else-expand, (origin: "c")) }
            { (>>[d ,], policy: fit-else-expand, (origin: "d")) }
            { (>>[e )], policy: fit-else-expand, (origin: "e")) }
          }
          { (>>[+ c ) ;], policy: uninitialized) }
        }
      }
```

For `a(b,c)+d(e,f)`, I'd lean towards attaching `+` to its left.  If you feel this change set is getting too large, you may handle it separately as a follow-up, I'll let you make that call.  It's ok to avoid handling binary expressions for now, and just focus on the partitioning around ','.